### DISCUSSION TITLE
What versions of OpenSSL3 are FIPS certified?

### DISCUSSION THREAD

**Discussion by Bonn93** [openssl/openssl#21552] [Q&A] [answered]
I am looking to use OpenSSL3 with FIPS and have been compiling some versions and following the FIPS install and config options. But I am unable to find exactly which versions are "certified for FIPS". 

The FIPS certificate states 3.0.0 is certified here https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4282

Does this mean that 3.0.1/2/3 etc are not certified? 

Thanks :)

**Comment by paulidale** [comment]
The FIPS provider from OpenSSL 3.0.0 is certified.  That's all currently.

This provider can be used with any OpenSSL 3.0 release.

So, build OpenSSL 3.0.0 to get the FIPS provider.  Build 3.0.8 without FIPS and then use the 3.0.0 FIPS provider with it.

**Comment by Bonn93** [comment]
I understand the concept, this makes sense but im not too sure on what exactly the provider is, could you provide an example of building to get the provider?

**Comment by paulidale** [comment]
Configure, build and install OpenSSL 3.0.0 exactly as per the [security policy](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf).

Then configure, build and install OpenSSL 3.0.8 without the `enable-fips` option.

**Comment by slontis** [comment]
Examples of provider related files are fips.so & legacy.so
libcrypto.so includes inbuilt providers for default and base.

So Use 3.0.0 code to produce the fips.so, and then use 3.0.X to build libcrypto and legacy.

**Comment by Jayukie** [comment]
I found that some patches of FIPS 140-3 have been included in openssl 3.1. 
Can I build OpenSSL 3.1 to get the FIPS provider, and build 3.0.4 without FIPS, and then use the 3.1 FIPS provider with it ?

**Comment by paulidale** [comment]
No, 3.1 is not validated.  It's not even been submitted for validation at this stage.

**Comment by Bonn93** [comment]
Hey, I'd like to re-open this, the docs now suggest otherwise, and the CMVP/Cert issued links 3.0.8, is 3.0.0 and 3.0.8 valid?

**Comment by t8m** [comment]
The 3.0.0 version of the fips provider was the first one FIPS validated.

Then the validated version was updated to 3.0.8 as is currently in the security policy.

So technically speaking currently only 3.0.8 is validated although 3.0.0 was originally validated too. We have an update of the security policy in progress that is supposed to make 3.0.0 validated again.

**Comment by paulidale** [comment, marked as answer]
The documents are a moving target.  They change over time as NIST changes the rules and transitions algorithms.  You are expected to keep track of and adjust to the changes.

When we validated 3.0.8, 3.0.0 was dropped from the certificate by mistake.  We're currently undergoing the process required to put it back again.  Once 3.0.0 is back on the certificate, there will be additional restrictions as to its use because of CVEs and NIST transitions.  Somethings that we permitted are not allowed anymore and your applications must stop doing them.

**Reply by Bonn93** [reply]
Thanks for the clarification. It's confused a few folks, is there a way to follow these ie blogs from openssl or something? 

Thanks again, much appreciated folks!

**Reply by paulidale** [reply]
We don't have anything specific.  Something to look into I suspect.

The closest we've got is the table on the [download page](https://www.openssl.org/source/) which points to the certificate and relevant security policy.

**Reply by mspncp** [reply]
@paulidale it would be great if the project would also indicate upcoming FIPS-validations. AFAIK, meanwhile some version 3.1.x is currently undergoing evaluation, but I don't know which one.

This could happen either on the download page, or elsewher if you know a more suitable location.

> The following OpenSSL version(s) are FIPS validated: 
> ...
>The following OpenSSL version(s) are currently undergoing FIPS evaluation:
> ...

**Comment by t8m** [comment]
> @paulidale it would be great if the project would also indicate upcoming FIPS-validations. AFAIK, meanwhile some version 3.1.x is currently undergoing evaluation, but I don't know which one.

The problem is that version can change at any time before the certificate is actually awarded.

**Reply by mspncp** [reply]
Ah, I understand. But is there at least a rough estimate when the validation can be expected?

**Reply by t8m** [reply]
Rough estimate = 2024 :grin:

**Reply by paulidale** [reply]
And even 2024 could be missed.
The 3.1 submission has been stalled for a while now :cry:
